File: frontend/lib/entities/medication_image_url_entity.dart
// File Name: medication_image_url_entity.dart
// Role: Defines the client trust policy for remotely loaded medication images.

const Set<String> _trustedMedicationImageHosts = {'nedrug.mfds.go.kr'};

// Function Name: safeMedicationImageUrl
// Description:
// - Returns a medication image URL only when it uses HTTPS and an approved
//   public-data host.
// - Rejects credentials, alternate ports, local endpoints, and arbitrary
//   third-party tracking hosts before Flutter can open a network connection.
// Parameters:
// - value: The untrusted URL value received from an API or persisted record.
// Returns:
// - The trimmed trusted URL, or an empty string when validation fails.
String safeMedicationImageUrl(dynamic value) {
  final text = value?.toString().trim() ?? '';
  if (text.isEmpty || text.length > 3000) {
    return '';
  }

  final uri = Uri.tryParse(text);
  if (uri == null ||
      uri.scheme.toLowerCase() != 'https' ||
      !uri.hasAuthority ||
      uri.host.isEmpty ||
      uri.userInfo.isNotEmpty ||
      (uri.hasPort && uri.port != 443)) {
    return '';
  }

  final host = uri.host.toLowerCase().replaceFirst(RegExp(r'\.$'), '');
  if (!_trustedMedicationImageHosts.contains(host)) {
    return '';
  }
  return text;
}

File: backend/entities/medication_image_url_entity.py
# File Name: medication_image_url_entity.py
# Role: Defines the backend trust policy for remotely loaded medication images.

from urllib.parse import urlsplit


TRUSTED_MEDICATION_IMAGE_HOSTS = frozenset({"nedrug.mfds.go.kr"})
MAX_MEDICATION_IMAGE_URL_LENGTH = 2_048


# Function Name: safe_medication_image_url
# Description:
# - Normalizes a public medication image URL and returns it only when it uses
#   HTTPS on the documented MFDS image host.
# - Rejects credentials, alternate ports, local endpoints, and arbitrary
#   third-party tracking hosts before the value is stored or returned.
# Parameters:
# - value: Untrusted URL value received from a client, catalog, or database.
# Returns:
# - The trusted normalized URL, or an empty string when validation fails.
def safe_medication_image_url(value: object) -> str:
    text = str(value).strip() if value is not None else ""
    if text.startswith("//"):
        text = f"https:{text}"
    if not text or len(text) > MAX_MEDICATION_IMAGE_URL_LENGTH:
        return ""

    try:
        parsed_url = urlsplit(text)
        hostname = (parsed_url.hostname or "").rstrip(".").casefold()
        port = parsed_url.port
    except ValueError:
        return ""

    if (
        parsed_url.scheme.casefold() != "https"
        or not hostname
        or parsed_url.username is not None
        or parsed_url.password is not None
        or (port is not None and port != 443)
        or hostname not in TRUSTED_MEDICATION_IMAGE_HOSTS
    ):
        return ""
    return text
